File: apps/android/app/src/main/java/com/dndsync/android/sync/CloudEnvelopeCodec.kt
package com.dndsync.android.sync

import android.util.Base64
import com.dndsync.proto.v1.CloudEnvelope

object CloudEnvelopeCodec {
    const val PAYLOAD_DND_STATE = Wire.PAYLOAD_DND_STATE
    const val PAYLOAD_PAIR_CONTROL = Wire.PAYLOAD_PAIR_CONTROL

    fun make(
        pairId: String,
        sender: String,
        ciphertext: ByteArray,
        payloadKind: Int = PAYLOAD_DND_STATE,
    ): CloudEnvelope =
        CloudEnvelope.newBuilder()
            .setVersion(Wire.PROTO_VERSION)
            .setPairId(pairId)
            .setSender(sender)
            .setCiphertext(com.google.protobuf.ByteString.copyFrom(ciphertext))
            .setPayloadKind(payloadKind)
            .build()

    fun decode(bytes: ByteArray): CloudEnvelope? {
        val envelope = try {
            CloudEnvelope.parseFrom(bytes)
        } catch (_: Exception) {
            return null
        }
        if (envelope.version != Wire.PROTO_VERSION) {
            return null
        }
        return envelope
    }

    fun decodeBase64(value: String): CloudEnvelope? {
        val bytes = try {
            Base64.decode(value, Base64.DEFAULT)
        } catch (_: IllegalArgumentException) {
            return null
        }
        return decode(bytes)
    }
}

File: apps/macos/FocusSync/CloudEnvelopeCodec.swift
import Foundation

enum CloudEnvelopeCodec {
    static let protoVersion: UInt32 = 1
    static let maxPushBytes = 4 * 1024
    static let payloadDndState: UInt32 = 0
    static let payloadPairControl: UInt32 = 1

    static func make(
        pairId: String,
        sender: String,
        ciphertext: Data,
        payloadKind: UInt32 = payloadDndState
    ) -> Dndsync_V1_CloudEnvelope {
        var envelope = Dndsync_V1_CloudEnvelope()
        envelope.version = protoVersion
        envelope.pairID = pairId
        envelope.sender = sender
        envelope.ciphertext = ciphertext
        envelope.payloadKind = payloadKind
        return envelope
    }

    static func encode(_ envelope: Dndsync_V1_CloudEnvelope) throws -> Data {
        try envelope.serializedData()
    }

    static func decode(_ data: Data) -> Dndsync_V1_CloudEnvelope? {
        guard let envelope = try? Dndsync_V1_CloudEnvelope(serializedBytes: data) else {
            return nil
        }
        guard envelope.version == protoVersion else {
            return nil
        }
        return envelope
    }

    static func decodeBase64(_ value: String) -> Dndsync_V1_CloudEnvelope? {
        guard let data = Data(base64Encoded: value) else {
            return nil
        }
        return decode(data)
    }
}
